Add EIP-5748: Approval Expiration for EIP-20 Tokens (#5748)
ethereum/EIPs: EIPS/eip-5748.md

## Rationale

`_expiration` is implemented as a mapping to improve the compatibility of the code.

A separate approval method with the header `approve(address spender, uint256 amount, uint256 period)` has been introduced to allow users to customize the `_expiration` value of the approval.

The internal method `_spendAllowance()` is introduced for code cleanliness with the function of checking the allowance amount and expiration.

The `allowance()` method has been modified to accommodate the functionality described in this standard. `allowanceExpiration()` has been added query the allowance expiration date.

## Backwards Compatibility

This standard is compatible with the [EIP-20](./eip-20.md), [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md) standards. Tokens issued in the form of proxy contracts can be updated to comply with this standard.

So that existing event statistics are not perturbed by the implementation of this standard, `_approve()` is triggered by the new event `AllowanceExpireTimeUpdated()`.

> This standard is absolutely compatible with the traditional [EIP-20](./eip-20.md) standard. `DEFAULT_EXPIRATION` is added so that the original `approve(address spender, uint256 amount)` method header can remain for backwards compatibility and ease of programming. `DEFAULT_EXPIRATION` can be set to a constant or variable according to the developer's needs.

Reviewer: Can we know this is 'absolutely compatible' ?

---
eip: 5748

> title: Approval Expiration for ERC-20 Tokens

Reviewer 2: ```suggestion
title: Approval Expiration for EIP-20 Tokens
```

---
eip: 5748
title: Approval Expiration for EIP-20 Tokens

> description: This standard introduces a standardized way for users to set expiration limits on ERC-20 token approvals.

Reviewer 3: ```suggestion
description: Extending EIP-20 approvals to automatically expire after a duration
```

author: Go+ Security (@GoPlusSecurity), Jeff Hall (@Mr-Lucky), Xavi (@XaaaaavitheFool), Turan Vural (@turanzv)

> discussions-to: <https://ethereum-magicians.org/t/eip-5748-automatic-expiration-of-token-approvals/>

Reviewer 3: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-5748-automatic-expiration-of-token-approvals/
```

> discussions-to: <https://ethereum-magicians.org/t/eip-5748-automatic-expiration-of-token-approvals/>

Reviewer 3: This link doesn't seem to exist.

Reviewer 4: 👀🐞🧑‍🔧✅

status: Draft
type: Standards Track
category: Interface

> created: 2022-10-02) format

Reviewer 3: ```suggestion
created: 2022-10-02
```

> requires:

Reviewer 3: This line is breaking the validator :rofl: I think you probably mean to depend on EIP-20 anyway:

```suggestion
requires: 20
```

Reviewer 4: 🙏🏼🙏🏼🙏🏼

---

> This standard introduces a standard way for users to set expiration limits on ERC-20 token approvals to prevent the loss of assets due to contractrual vulnerabilities or authorization attacks.

Reviewer 3: ```suggestion
```

## Abstract

> This standard adds an expiration time to the `_approve()` function of token contracts to automatically recall approvals within a certain time period. This functionality as a standard will prevent vulnerabilty exploits due to outstanding token approvals.

Reviewer 3: This proposal doesn't add an expiration time, but rather a **validity period** after the transaction is included on-chain. I know it's a pedantic difference, but it's important from a security point of view.

In the Transit Swap attack of October 1st, 2022, hackers used a vulnerability in the Transit Swap contract to steal $20 million in assets. Due in part to professional security audits and the strong brand recognition of Transit Swap, many users were given a false sense of security and trusted the platform's service. Transit Swaps's contract had been running smoothly for over a year and had accumulated a large amount of user allownces, making the contract a high-reward target for malicious actors.

There have been many similar incidents to the Transit Swap attack. Security audits cannot gurantee that a contract is 100% free from vulnerabilities. Some vulnerabilities are intentionally left by malicious developers themselves. In line with good trust practices and general security hygeine, unconditional trust is risky, and modifying approvals into limited trusts with automatic recovery will be a key tool to natively prevent attacks.

> 

Reviewer 3: This motivation is good background (though it might be polite to avoid mentioning a specific company/product), but it doesn't actually convince me that an expiring approval solves these problems. Maybe add a sentence along the lines of:

```markdown
If these lingering approvals had instead expired, the number of users affected by this attack would have been much smaller.
```

## Specification

> A time period `_expiration` is mapped on top of the original `allowance` to specify an approval time limit.

Reviewer 3: What does "mapped on top of" mean here?

> The internal  `_approve()` method adds a uint256 `period` to the entry and stored in `_expiration`.

Reviewer 3: EIP-20 does not define an `_approve()` method. If you are referring to a particular implementation of EIP-20, don't. Instead, limit your specification to the visible interface of the contract and not any implementation details.

Reviewer 4: seen, edited

> A `DEFAULT_EXPIRATION` constant is added for implementations that do not specify the `period` parameter. A getter method `defaultExpiration()` is included in the implementation for transparency to the user in the approval process.

Reviewer 3: Your tone here is a bit... off. You're writing your proposal like you're walking someone through your decision process. While there is a place for this type of information (the `Rationale` section), the `Specification` section should clearly state the requirements for a contract to meet this standard. Perhaps something like:

```suggestion
Contracts implementing this standard MUST expose a `DEFAULT_EXPIRATION` constant of type `uint256`. This constant stores the expiration period to use for `approve` calls that do not include a `period` argument.

`approve` functions without a `period` argument MUST use the expiry period exposed in the `DEFAULT_EXPIRATION` constant.
```

> In order not to interfere with the existing event statistics and to allow other developers to count the validity of allowance, we add a new event `AllowanceExpireTimeUpdated()`, which will be triggered in the `_approve()` method.

Reviewer 3: Again, the tone is off. You should state explicitly when compliant contracts emit the `AllowanceExpireTimeUpdated` event.

A separate approval method with the header `approve(address spender, uint256 amount, uint256 period)` modifies the `_expiration` value for existing approvals.

> ```solidity

Reviewer 5: Move to Reference Implementation if you didn't intend to specify a specific behavior but only the interface.

> contract ERC20 is Context, IERC20, IERC20Metadata {

Reviewer 3: Instead of using a contract, please only include an interface. EIPs should only specify the visible portion of a contract, and leave the implementation up to implementers.

        require(currentAllowance >= subtractedValue, "ERC20: decreased allowance below zero");
        unchecked {
            _approve(owner, spender, currentAllowance - subtractedValue, DEFAULT_EXPIRATION);
        }

        return true;
    }

    function decreaseAllowance(address spender, uint256 subtractedValue, uint256 period) public virtual returns (bool) {
        address owner = _msgSender();
        uint256 currentAllowance = allowance(owner, spender);
        require(currentAllowance >= subtractedValue, "ERC20: decreased allowance below zero");
        unchecked {
            _approve(owner, spender, currentAllowance - subtractedValue, period);
        }

        return true;
    }
```

The above modifications adds a validity period to ERC-20 token approvals. Users can freely update their allowance validity at any time by calling `approve(address spender, uint256 amount, uint256 period)`.

## Rationale
`_expiration` is implemented as a mapping to improve the compatibility of the code.

So that existing event statistics are not perturbed by the implementation of this standard, `_approve()` is triggered by the new event `AllowanceExpireTimeUpdated()`.

> This standard is absolutely compatible with the traditional ERC-20 standard.  `DEFAULT_EXPIRATION` is added so that the original `approve(address spender, uint256 amount)` method header can remain for backwards compatibility and ease of programming. `DEFAULT_EXPIRATION` can be set to a constant or variable according to the developer's needs.

Reviewer 3: Please move this paragraph to the Backwards Compatibility section.

Reviewer 4: committed

---
eip: 5748
title: Approval Expiration for EIP-20 Tokens
description: Extending EIP-20 approvals to automatically expire after a duration
author: Go+ Security (@GoPlusSecurity), Eskil (@xyfidea), Jeff Hall (@Mr-Lucky), Xavi (@XaaaaavitheFool), Turan Vural (@turanzv)
discussions-to: https://ethereum-magicians.org/t/eip-5748-automatic-token-approval-expiration/11185
status: Draft
type: Standards Track
category: Interface
created: 2022-10-02
requires: 20
---

## Abstract

> This standard adds an expiration time to the `_approve()` function of token contracts to automatically recall approvals within a certain time period. This functionality as a standard will prevent vulnerabilty exploits due to outstanding token approvals.

Reviewer 3: I suspect `_approve` comes from a well known EIP-721 implementation? The standard itself doesn't define it.
